src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs

```diff
@@ -2436,118 +2477,321 @@ public void UpdateIsNullableEnabledForConstructorsAndFields(bool useStatic, bool
+                {
+                    builder.Add(group.ToImmutableAndFree());
                 }
 
-                group.Free();
+                initializers.Free();
                 return builder.ToImmutableAndFree();
             }
 
             public void Free()
             {
-                NonTypeNonIndexerMembers.Free();
-                StaticInitializers.Free();
-                InstanceInitializers.Free();
-                IndexerDeclarations.Free();
+                NonTypeNonIndexerMembers?.Free();
+                InstanceInitializersForPositionalMembers?.Free();
             }
+        }
 
-            internal void AddOrWrapTupleMembers(SourceMemberContainerTypeSymbol type)
+        protected virtual MembersAndInitializers? BuildMembersAndInitializers(DiagnosticBag diagnostics)
+        {
+            var declaredMembersAndInitializers = getDeclaredMembersAndInitializers();
+            if (declaredMembersAndInitializers is null)
             {
-                this.NonTypeNonIndexerMembers = type.AddOrWrapTupleMembers(this.NonTypeNonIndexerMembers.ToImmutableAndFree());
+                // Another thread completed the work before this one
+                return null;
+            }
+
+            var membersAndInitializersBuilder = new MembersAndInitializersBuilder(declaredMembersAndInitializers);
+            AddSynthesizedMembers(membersAndInitializersBuilder, declaredMembersAndInitializers, diagnostics);
+
+            if (_lazyMembersAndInitializers != null)
+            {
+                // Another thread completed the work before this one
+                membersAndInitializersBuilder.Free();
+                return null;
+            }
+
+            return membersAndInitializersBuilder.ToReadOnlyAndFree(declaredMembersAndInitializers);
+
+            DeclaredMembersAndInitializers? getDeclaredMembersAndInitializers()
+            {
+                var declaredMembersAndInitializers = _lazyDeclaredMembersAndInitializers;
+                if (declaredMembersAndInitializers != null)
+                {
+                    return declaredMembersAndInitializers;
+                }
+
+                if (_lazyMembersAndInitializers is not null)
+                {
+                    // We're previously computed declared members and already cleared them out
+                    // No need to compute them again
+                    return null;
+                }
+
+                var diagnostics = DiagnosticBag.GetInstance();
+                declaredMembersAndInitializers = buildDeclaredMembersAndInitializers(diagnostics);
+
+                var alreadyKnown = Interlocked.CompareExchange(ref _lazyDeclaredMembersAndInitializers, declaredMembersAndInitializers, null);
+                if (alreadyKnown is not null)
+                {
+                    diagnostics.Free();
+                    return alreadyKnown;
+                }
+
+                AddDeclarationDiagnostics(diagnostics);
```

Comment:
>AddDeclarationDiagnostics(diagnostics); [](start = 16, length = 39)

It looks like there is a race around this call. I think it can be mitigated as follows:
1) Instead of reporting diagnostics on the spot, we add it to DeclaredMembersAndInitializers as an immutable array
2) Then as part of ```membersAndInitializersBuilder.ToReadOnlyAndFree(declaredMembersAndInitializers)``` call above, we append that array to ```diagnostics``` which is a parameter of the enclosing function (BuildMembersAndInitializers).

Reply:
That seems to be the case (duplicate diagnostics reported in some tests), but I don't understand why.
Only one thread should come back from `Interlocked.CompareExchange` with `alreadyKnown` being `null`, no?

---
In reply to: [562995947](https://github.com/dotnet/roslyn/pull/50650#discussion_r562995947) [](ancestors = 562995947)

Reply:
I see. The problem is caused by resetting `_lazyDeclaredMembersAndInitializers` (your other comment clarified, thanks)

---
In reply to: [562999875](https://github.com/dotnet/roslyn/pull/50650#discussion_r562999875) [](ancestors = 562999875,562995947)